### ISSUE TITLE
requests from grpc-js on bidi channel sometimes get multiple-seconds delay before being sent

### ISSUE THREAD

**Issue by kskalski** [grpc/grpc-node#882]
### Problem description
I have a gRPC .NET server and grpc / grpc-js client (in Electron main process) connecting over loopback (with client ssl certificates enabled). Client starts a bidi stream and occasionally receives messages from server or sends requests to it.
I just switched from grpc implementation of the client to grpc-js one, but then I noticed that requests from client to server get occasionally delivered with significant delay (2-10s), which doesn't happen at all when using grpc implementation.

### Reproduction steps
The messages and channel outline is like that:

```
message WindowFormsToExternal {
  string window = 1;
  string action = 2;
  repeated string args = 3;

  int32 ExternalSequenceNr = 4;
  int32 FormsSequenceNr = 5;
}
message WindowExternalToForms {
  string window = 1;
  string action = 2;
  repeated string args = 3;

  int32 ExternalSequenceNr = 4;
  int32 FormsSequenceNr = 5;
}

service ElectorGateway {
  rpc WindowControl(stream WindowExternalToForms) returns (stream WindowFormsToExternal) }
```

The server and client implementations are within larger app, I can try to create a self-contained repro if this is issue cannot be easily tracked down.

I create the client by using
```
var ssl_creds = grpc.credentials.createSsl(creds.ca_cert, creds.ui_key, creds.ui_cert);
var elector_proto = grpc.loadPackageDefinition(packageDefinition)['pro']['elector']['proto'];
this.grpc_client = new elector_proto.ElectorGateway('127.0.0.1:15745', ssl_creds,
  { 'grpc.ssl_target_name_override': 'ElectorAppGen', 'grpc.default_authority': 'ElectorAppGen'});
```

### Environment
 - Windows 10, May 2019 update
 - Electron 5.0.1 on the client, .NET 4.6.1 on the server
 - grpc-js 0.4.0 on the client (comparing with grpc 1.20.3), gRPC .NET 1.21.0 on the server

### Additional context
Thanks for making grpc-js and grpc seemlessly replacable, thanks to that I can clearly pin the problem to grpc-js, since with single-line change I'm testing both variants that the issue only happens when using grpc-js

**Comment by kskalski** [comment]
I debugged this further and it is related to the context from which the message is sent. The multi-seconds latencies occur when message is pushed from Electron's IPC engine handler, e.g.
```
win.on("minimize", (event) => {
   console.log('sending');
    this.ui_communicator.SendMessage({ window: "main", action: Date.now().toString() }, (resp) => dialog.showMessageBox(resp));
});
```

I have a full repro here: https://github.com/kskalski/snippets/tree/master/GrpcJsElectronSharp
- change paths to node, npm, MSBuild in build.sh
- ``./build.sh``
- ``./Server/bin/Debug/net461/Server.exe``  in one terminal
- ``./Client/win-ia32-unpacked/Client.exe`` in another terminal
- minimize / maximize the Electron window (fires send message code)
- observe message arriving to server after many seconds

Contrast it with sending messages from setInterval (in Client's main.ts) or by using 'grpc' instead of '@grpc/grpc-js' in comm.ts, which doesn't suffer from latency problem.

**Comment by kskalski** [comment]
Not sure then if this is node/Electron's bug or grpc-js', but a reasonable workaround is to wrap the message sending with ``setImmediate``. The multi-second latencies get reduced to 2-10ms

**Comment by murgatroid99** [comment]
Thank you for providing a reproduction case, but the fact that it is using `pbts` to generate code is making it hard to follow how your code interacts with grpc. We do not officially support that code generation method and I don't know what it does. Also, do I understand correctly that

**Comment by kskalski** [comment]
Actually the pbts-generated types are practically not used in the repro code... I could just remove it if it makes it easier, the code defining client and sending messages is only replying on proto-loader

**Comment by murgatroid99** [comment]
I think that would help me follow along with what the code is doing.

**Comment by kskalski** [comment]
I pushed the new version.

**Comment by kskalski** [comment]
@theogravity did you try wrapping message send calls on the client in ``setImmediate``? I still use that workaround and it's effective for me. Not sure if that is related though, because in my case it's clearly dependent on JavaScript execution context (node/Electron's background executor?) and happens on a loopback path even with single requests, while your scenario seems to happen only under load.

**Comment by kskalski** [comment]
> Are you saying do something like?
> 
Yes, exactly

**Comment by eulertour** [comment]
I'm running into this problem as well, where the `setImmediate` workaround is mitigating but not preventing the problem. Without `setImmediate` the RPCs hang immediately and with it they hang after ~50 RPCs. A fix would be greatly appreciated.
